pkg/config/default.go

```diff
@@ -188,6 +189,25 @@ const (
 	DefaultVolumePluginTimeout = 5
 )
 
+func defaultSigPath() (string, error) {
+	// NOTE: For now we want Windows to use system locations.
+	// GetRootlessUID == -1 on Windows, so exclude negative range
+	if unshare.GetRootlessUID() > 0 {
+		configHome, err := homedir.GetConfigHome()
```

Comment:
I don't see how this fixes https://github.com/containers/podman/issues/23818
Without an extensive code analysis I would assume there are more code paths calling into this

I would think the proper fix is to make `homedir.GetConfigHome()` not error on ENOENT IMO

Reply:
I disagree, Callers might want to make different decisions on no ConfigHome and having to look at "", nil versus "", ENOENT allows them to fail if they expect to have a ConfigHome.

Reply:
It should fix the specific error, and seems like a good change in general, since it is expected that the SigPath does not exist in the users HomeDir is common.

Reply:
Why would a caller expect this this path to exists? For pretty much all other tools this path is used to read option config files. I do not even understand why thins function tries to create the given path.
IMO this should return the path,nil even if the underlying path does not actually exists.

If you call something like os.UserHomeDir() you also juts the the path but no guarantee that it exists.

Reply:
Ok you could change the function, but this would be potentially breaking change for callers of the function.

Reply:
> Without an extensive code analysis I would assume there are more code paths calling into this

There are several, incl. things like `pkg/machine/env.GetConfDir` which uses the return value to maintain state. We can’t just _silently ignore_ failures without inventing some other place where to store data. And if we were to invent some other place, in a situation where `~/.config` both doesn’t exist and can’t be created … I can’t imagine what that would be.

We _must_ fail in this situation, ultimately.

---

> Callers might want to make different decisions on no ConfigHome and having to look at "", nil versus "", ENOENT allows them to fail if they expect to have a ConfigHome.

(I can’t parse this, but, anyway):

Please make a table of:
- on-filesystem situation
- environment setting
- expected behavior

After we have the expected behavior, we can discuss what the API should look like.

I _generally_ think that it is _undesirable_ for callers to make any kinds of individual decisions based on the situation. All callers who wanted “$XDG_CONFIG_HOME or a fallback” should get _exactly the same_ “$XDG_CONFIG_HOME or a fallback” and generally behave the same in the fallback situation. It seems to me, as a first approximation, undesirable to expose the details of the fallback to the callers; instead, the fallback logic should all be inside `GetConfigHome`.

[Or, to put it another way, if some callers don’t want the standard $XDG_CONFIG_HOME behavior, they shouldn’t be using the $XDG_CONFIG_HOME variable and name at all; they should define some other variable and name for that custom behavior.]

---

… personally, I suggest that we don’t want all of this complexity, and “each Linux user must have a modifiable $ConfigHome” is a reasonable thing which we just assume, and fail otherwise.

Especially consider the case where `GetConfigHome` is intentionally returning an error, where $XDG_CONFIG_HOME is _set to a value_ but that value is not writeable by the current user (typically using `su` without `--login`). That’s an unreasonable, potentially dangerous, setup, and we should _not_ be triggering any fallback behavior; we should fail at the first opportunity.

Reply:
https://github.com/containers/podman/issues/23818#issuecomment-2328296340

> Especially consider the case where GetConfigHome is intentionally returning an error, where $XDG_CONFIG_HOME is set to a value but that value is not writeable by the current user (typically using su without --login). That’s an unreasonable, potentially dangerous, setup, and we should not be triggering any fallback behavior; we should fail at the first opportunity.

Well yeah but do we have to fail when the files are never needed in the first place? I would assume code fails anyway if we get EACCESS when reading the config files as we only ignore ENOENT AFAIK.

The doc comments on GetConfigHome() make no such claim that they check for the dir existence, in fact when XDG_CONFIG_HOME is set it doesn't try to create/stat it there so your described scenario isn't failing there.

Reply:
Added https://github.com/containers/storage/pull/2084 to handle XDG_CONFIG_HOME versus $HOME/.config

Reply:
> Well yeah but do we have to fail when the files are never needed in the first place?

I think that’s an very reasonable point when looking for general config files.

When looking for security-relevant configuration like `policy.json`, it gets… complicated, because it’s possible the user caused this situation by a mistaken chmod, and falling back to the system-wide config might be a much more permissive policy and not be what the user wanted. Now, to an extent, that’s just inherent in the “check if a per-user config exists, otherwise fall back” idea — we can protect the user against _some_ mistakes but we can’t protect against e.g. a typo in the config file name.